Add repo list

olearycrew opened the pull request:
**Description**
This change adds a command `glab repo list` that will list the user's owned repos (projects) by any chosen order / sort like other lists - similar to and modeled after the `glab repo contributors` list.

**Related Issue**
Resolves #671

**How Has This Been Tested?**
I tested this with my own account, using various orders and sort directions.

**Screenshots (if appropriate):**
![image](https://user-images.githubusercontent.com/6044920/114966151-ff66fd80-9e3f-11eb-85f7-968a73a1ab40.png)

**Types of changes**
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation
- [ ] Chore (Related to CI or Packaging to platforms)

profclems committed:
feat: add `release create` command

profclems committed:
feat: add `release create` command

zemzale committed:
fix(main): Allow host override with full URLs

When trying to override host with full URLs, config key lookups would
fail. This came down to the fact that we use `my.gitlab.com` as config
key and when providing override as `https://my.gitlab.com/` it would
fail to find this config key.

Issue #592

Commit:
feat(ci/status): list of external jobs and fix when no jobs are found for the pipeline

huegelcesar committed:
feat(commands/api): add a reviewer filter to the mr list command

Delivers #680

Commit:
Add repo list

Commit:
Fix title

jeanphi-baconnais commented:
Hi @olearycrew , i've check your PR and it's seems to be OK 👍

zemzale commented on commands/project/list/list.go:
```diff
@@ -0,0 +1,95 @@
+
+	l := &gitlab.ListProjectsOptions{
+		OrderBy:        gitlab.String(opts.OrderBy),
+		MinAccessLevel: gitlab.AccessLevel(50), // Only projects you own
```
I think it would be better to set this to Minimal access instead?

I tried to run this against my GitLab.com account. That showed me most of my projects(there is a [bug](https://gitlab.com/gitlab-org/gitlab/-/issues/219299)).

Then I tried to run against one of my self hosted instances, and it showed me nothing. When I go into Web UI I have 5 projects available, and I expected to see the same. The difference comes from the fact that I am just a developer on these projects.

And I feel that a lot of people might have such instances where they are only developers on a lot of projects and would want to see them, but can't.

profclems replied:
Yeah it tried it too on GitLab.com and it only show repos I own without including projects in my personal namespace

```
$ glab repo list

Showing 2 of 2 projects (Page 1 of 1)

glab-cli/test     
glab-cli/test-pv  
```

I suggest instead of restricting the result with the min access level, we should have flags to allow the user get the type of repos to return
```go
repoListCmd.Flags().BoolVarP(&opts.All, "all", "a", false, "list all projects on the instance")
repoListCmd.Flags().BoolVarP(&opts.Owned, "mine", "m", false, "limit by projects you own")
repoListCmd.Flags().BoolVar(&opts.Member, "member", false, "limit by projects which you are a member")
repoListCmd.Flags().BoolVar(&opts.Starred, "starred", false, "Get starred projects")
```

zemzale reviewed (changes requested):
I looked over the PR and have one slight issue with what projects are actually being shown.

Other than that this looks fantastic.

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.2.0 to 2.2.1

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.2.0 to 2.2.1.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.2.0...v2.2.1)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/AlecAivazis/survey/v2 from 2.2.9 to 2.2.12

Bumps [github.com/AlecAivazis/survey/v2](https://github.com/AlecAivazis/survey) from 2.2.9 to 2.2.12.
- [Release notes](https://github.com/AlecAivazis/survey/releases)
- [Commits](https://github.com/AlecAivazis/survey/compare/v2.2.9...v2.2.12)

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
: CI status now has an option to show a more compact view.

profclems committed:
Fix method not allowed when merging MR

When merging MR, the --when-pipeline-succeeds flag is by default set to true.
Due to that, if a project does not have any pipeline set, results in error
`method not allowed`.
This commit fixes this issue by checking if the merge request has a pipeline set.
Also, on TTYs, users are prompted to confirm whether to merge when pipeline succeeds or not,
if the MR has a pipeline set.

profclems committed:
Fix segfault on ci list

For gitlab instances running on versions <= 11, the pipeline list api does not return the `created_at` field causing `glab` to panic.
This commit checks if the created_at field is available before using it

Reolves #688

kousikmitra committed:
feat(mr/view): show reviewers of the merge request

Delivers #692

kousikmitra committed:
feat(mr/view): update mr reviewers

It allowes `mr update` to update reviewers with a new flag --reviewers
Delivers #562

profclems committed:
feat: mr merge options

This adds merge options when merging an MR. The user is given three merge options
- create merge commit
- squash and merge
- rebase and merge

NB: This options are only available on TTYs. On non-ttys, the default is to create a merge commit unless the user explicitly specifies the merge method using the available flag options

profclems committed:
fix reviewer action messages

profclems committed:
Allow protocol override when overriding the default host

This allows overriding the default protocol of the hostname if a fully qualified url is provided when using the `GITLAB_HOST` environment variable.
For example, if the GITLAB_HOST=http://my.glinstance.org, then the host will be set to my.glinstance.org and the protocol become http, unlike before when the protocol was ignored and only the hostname was considered.

This is a follow-up for #678
and #667

dependabot[bot] committed:
chore(deps): bump github.com/xanzy/go-gitlab from 0.48.0 to 0.49.0

Bumps [github.com/xanzy/go-gitlab](https://github.com/xanzy/go-gitlab) from 0.48.0 to 0.49.0.
- [Release notes](https://github.com/xanzy/go-gitlab/releases)
- [Changelog](https://github.com/xanzy/go-gitlab/blob/master/releases_test.go)
- [Commits](https://github.com/xanzy/go-gitlab/compare/v0.48.0...v0.49.0)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/otiai10/copy from 1.5.1 to 1.6.0

Bumps [github.com/otiai10/copy](https://github.com/otiai10/copy) from 1.5.1 to 1.6.0.
- [Release notes](https://github.com/otiai10/copy/releases)
- [Commits](https://github.com/otiai10/copy/compare/v1.5.1...v1.6.0)

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
feat(repo): Create project from URL

This is gonna allow for creating projects using full URL which is gonna
override the host URL too.

Issue #618

Commit:
fix(glrepo): Remove side effect form RemoteURL

RemoteURL was changing one if it's passed in variables. This introduced
problems with subsequent calls to the function, because member
`RemoteArgs.Protocol` was changed to from `https` to `https://`. This
made every other call to the function skip over the protocol and just
return ssh URL.

Issue #710

profclems committed:
fix projectPathFromArgs

This ensures supoport for

- ssh urls
- urls with .git prefix
- respect hostname in url provided

profclems committed:
scriptability: Avoid truncating columns of tables/lists on non-TTYs

This is a hot fix to solve where glab still truncates lines with length longer that the specified column width on non-TTYs to improve scriptability

Resolves #706

profclems committed:
fix list tests

profclems committed:
ci status: improve scriptability

Ensures user is not prompted on non-ttys with options to retry or view logs

profclems committed:
fix race condition on mr merge

Merge fails when mr is rebasing and it is due to glab being too fast for GitLab. To solve this, we retry the merge api 3 times at a 6s interval each

Ref: https://github.com/profclems/glab/pull/695#issuecomment-829852012
Resolves #407

profclems committed:
Add -y|--yes flag to skip submission prompt

Resolves #697

profclems committed:
feat: git-credential store hack

profclems committed:
depend on git credentials for pull/push/clone commands

We don't need to use user credentials for pull/push/clone operations since the user can now opt to saving credentials to git-credential helper else, be prompted to login each time

profclems committed:
Do not prompt for api_protocol on default gitlab instance

maxice8 committed:
fix(commands/mr/merge): don't call retry.Opts from inside function

maxice8 committed:
fix(commands/mr/merge): avoid calling a value that can be nil after a failure

fixes #719

maxice8 committed:
feat(commands/mr/merge): don't retry merging if the error isn't unmergeable branch

we do this retry so we can work around a rebase causing the branch to
not be mergeable for a few seconds even though it has no conflicts.

so instead of retrying for every possible error we use `RetryIf` to
check if the error is the one expected and try again, otherwise we just
exit with the error

maxice8 committed:
feat(commands/mr/merge): if we are not rebasing before merging don't try multiple times

maxice8 committed:
refactor(commands/mr/merge): rewrite retry to be cleanier

maxice8 committed:
feat(api): return response when merging the Merge Request

this is required because we need to verify the status code that is
returned by GitLab

maxice8 committed:
fix(commands/mr/merge): check the status code to return a recoverable error

dependabot[bot] committed:
chore(deps): bump github.com/xanzy/go-gitlab from 0.49.0 to 0.50.0

Bumps [github.com/xanzy/go-gitlab](https://github.com/xanzy/go-gitlab) from 0.49.0 to 0.50.0.
- [Release notes](https://github.com/xanzy/go-gitlab/releases)
- [Changelog](https://github.com/xanzy/go-gitlab/blob/master/releases_test.go)
- [Commits](https://github.com/xanzy/go-gitlab/compare/v0.49.0...v0.50.0)

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
fix auth status tests

Commit:
docs(update): latest

profclems committed:
Update README.md

Commit:
fix job viewing with detached pipelines

For `ci trace` and `ci status`, the `GetPipelines` function has been
used. However, `GetPipelines` does not with [detached pipelines](https://docs.gitlab.com/ee/ci/merge_request_pipelines/).

To fix this, we're now using a new function `GetLastPipeline`, which
first tries to get the last pipeline through the `GetCommit` API and if
that should not return any results, it falls back to the old
`GetPipelines` function.

Some parts of the codes have been refactored to use this new function
instead of listing all pipelines and then grabbing the first elementt of
that list.

`GetPipelines` is now only used within the `api` package itself, but I
kept it public for compatibility reasons.

Commit:
Fix triggered jobs duration

This commit fixes a bug where a `glab ci status` would panic if there is
a job within that CI pipeline that is not triggered yet. In such a case,
the StartedAt field is not populated, which triggered a panic.

To fix this, we use `time.Now()` if `StartedAt` is nil, which means that
the job will be displayed with a duration of 0.

Fixes #727

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.2.1 to 2.3.1

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.2.1 to 2.3.1.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.2.1...v2.3.1)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.3.1 to 2.3.3

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.3.1 to 2.3.3.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.3.1...v2.3.3)

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
interactive options and better error handling

profclems committed:
fix test

deepsource-autofix[bot] committed:
improve empty string test

profclems committed:
feat: add `release upload` command

profclems committed:
feat: add `release delete` command

profclems committed:
fix markdown renderer color

profclems committed:
feat: `release view` command

profclems committed:
linting

profclems committed:
fix wrong refs

profclems committed:
fetch latest release instead of fetching tag first

profclems committed:
add footer notes to view

dependabot[bot] committed:
chore(deps): bump github.com/mattn/go-runewidth from 0.0.12 to 0.0.13

Bumps [github.com/mattn/go-runewidth](https://github.com/mattn/go-runewidth) from 0.0.12 to 0.0.13.
- [Release notes](https://github.com/mattn/go-runewidth/releases)
- [Commits](https://github.com/mattn/go-runewidth/compare/v0.0.12...v0.0.13)

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
more verbose error message

dependabot[bot] committed:
chore(deps): bump github.com/tidwall/pretty from 1.1.0 to 1.1.1

Bumps [github.com/tidwall/pretty](https://github.com/tidwall/pretty) from 1.1.0 to 1.1.1.
- [Release notes](https://github.com/tidwall/pretty/releases)
- [Commits](https://github.com/tidwall/pretty/compare/v1.1.0...v1.1.1)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/briandowns/spinner from 1.12.0 to 1.13.0

Bumps [github.com/briandowns/spinner](https://github.com/briandowns/spinner) from 1.12.0 to 1.13.0.
- [Release notes](https://github.com/briandowns/spinner/releases)
- [Commits](https://github.com/briandowns/spinner/compare/v1.12...v1.13.0)

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
feat: add `release download` command

profclems committed:
fix: `:id` placeholder in `api` command

dependabot[bot] committed:
chore(deps): bump github.com/mattn/go-isatty from 0.0.12 to 0.0.13

Bumps [github.com/mattn/go-isatty](https://github.com/mattn/go-isatty) from 0.0.12 to 0.0.13.
- [Release notes](https://github.com/mattn/go-isatty/releases)
- [Commits](https://github.com/mattn/go-isatty/compare/v0.0.12...v0.0.13)

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
fix `mr merge` failing on detached branch

`mr merge` fails with error `could not determine current branch: git: not on any branch` when run on a detached branch.

This is due to glab fetching the current branch first before checking if a merge request ID was provided as argument.

profclems committed:
fix hostname validator to accept IPs and hosts with port

Resolves #738

profclems committed:
add tests

profclems committed:
Update gitlab-issues-to-github.yml

profclems committed:
chore(packaging): glab docker images

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.3.3 to 2.3.5

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.3.3 to 2.3.5.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.3.3...v2.3.5)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.3.5 to 2.3.8

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.3.5 to 2.3.8.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.3.5...v2.3.8)

---
updated-dependencies:
- dependency-name: github.com/gdamore/tcell/v2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/briandowns/spinner from 1.13.0 to 1.15.0

Bumps [github.com/briandowns/spinner](https://github.com/briandowns/spinner) from 1.13.0 to 1.15.0.
- [Release notes](https://github.com/briandowns/spinner/releases)
- [Commits](https://github.com/briandowns/spinner/compare/v1.13.0...v1.15.0)

---
updated-dependencies:
- dependency-name: github.com/briandowns/spinner
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.3.8 to 2.3.9

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.3.8 to 2.3.9.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.3.8...v2.3.9)

---
updated-dependencies:
- dependency-name: github.com/gdamore/tcell/v2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
build/goreleaser: docker images

profclems committed:
ci/actions: modify release workflow

profclems committed:
ci/action: fix release workflow

profclems committed:
ci: auto add release news to github discussion

Commit:
docs(update): latest

dependabot[bot] committed:
chore(deps): bump github.com/tidwall/pretty from 1.1.1 to 1.2.0

Bumps [github.com/tidwall/pretty](https://github.com/tidwall/pretty) from 1.1.1 to 1.2.0.
- [Release notes](https://github.com/tidwall/pretty/releases)
- [Commits](https://github.com/tidwall/pretty/compare/v1.1.1...v1.2.0)

---
updated-dependencies:
- dependency-name: github.com/tidwall/pretty
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

profclems committed:
build: fix goreleaser

profclems committed:
ci/action: fix release workflow

profclems committed:
ci/action: fix release workflow

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.3.9 to 2.3.10

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.3.9 to 2.3.10.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.3.9...v2.3.10)

---
updated-dependencies:
- dependency-name: github.com/gdamore/tcell/v2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

jume-dev committed:
docs: Fix build instruction

jume-dev committed:
fix: Don't try to find branch if mrID is found

- ensures that mr operations also work on detached heads

dependabot[bot] committed:
chore(deps): bump github.com/gdamore/tcell/v2 from 2.3.10 to 2.3.11

Bumps [github.com/gdamore/tcell/v2](https://github.com/gdamore/tcell) from 2.3.10 to 2.3.11.
- [Release notes](https://github.com/gdamore/tcell/releases)
- [Changelog](https://github.com/gdamore/tcell/blob/master/CHANGESv2.adoc)
- [Commits](https://github.com/gdamore/tcell/compare/v2.3.10...v2.3.11)

---
updated-dependencies:
- dependency-name: github.com/gdamore/tcell/v2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Vendetta26 reviewed (commented):
Duplicate of #

Vendetta26 reviewed (commented):
Duplicate of #

profclems commented:
Closing in favour #906 which still has @olearycrew's commits
